fix(lte): Fixed `detect_init_system`

MoritzThomasHuebner opened the pull request:
Closes #14033


Signed-off-by: Moritz Huebner <moritz.huebner@tngtech.com>

<!--
    Tag your PR title with the components that it touches along with
    the type of change
    E.g. "fix(orc8r): Fix reindexer race condition" or "feat(agw) ..."
-->

## Summary

- Added `exec_command_capture_output` to allow running commands that return non-zero status codes without raising an exception.
- Moved `_is_installed` function outside `detect_init_system`.
- Simplified docker/systemd_magmad_running logic.
- Renamed `MagmadUtil._data` to `MagmadUtil._credentials` and removed "command" key.
- No longer write "command" to `MagmadUtil._credentials`, as this was not used anyway.


## Test Plan

See if an integration test can be started
- [x] with a dockerized setup
- [x] with a systemd setup 
<!--
    How did you test your change? How do you know it works?
    Add supporting screenshots, terminal pastes, etc. as necessary
-->

MoritzThomasHuebner committed:
fix(lte): Fixed `detect_init_system`

- Added `exec_command_capture_output` to have an option to capture stdout and continue on non-zero exit codes.
- Removed `_is_installed` function.
- Simplified docker/systemd_magmad_running logic.
- Renamed `MagmadUtil._data` to `MagmadUtil._credentials` and removed "command" key.
- No longer write "command" to `MagmadUtil._credentials`, as this was not used anyway.

Signed-off-by: Moritz Huebner <moritz.huebner@tngtech.com>

nstng reviewed (approved):
Thank you for fixing this - I was sure I checked the service not running case :( but yeah, I can reproduce the issue.

Refactoring also lgtm.

NIT I would have preferred a prior check if the tool is installed (was done via type before the change). It feels a little hacky that a 'command not found' is basically ignored and the empty stdout is compared in that case. Raw return of `subprocess.run`:
```
CompletedProcess(args=['sshpass', '-p', 'vagrant', 'ssh', '-o', 'UserKnownHostsFile=/dev/null', '-o', 'StrictHostKeyChecking=no', '-o', 'LogLevel=ERROR', 'vagrant@192.168.60.142', "docker ps --filter 'name=magmad' --format '{{.Names}}'"], returncode=127, stdout=b'', stderr=b'bash: docker: command not found\n')
```

MoritzThomasHuebner committed:
fix(lte): Added _is_installed check back in

Signed-off-by: Moritz Huebner <moritz.huebner@tngtech.com>

nstng commented on lte/gateway/python/integ_tests/s1aptests/s1ap_utils.py:
```diff
@@ -906,35 +903,43 @@ def exec_command_output(self, command):
+            "systemctl is-active magma@magmad",
+        )
+        docker_magmad_running = \
+            self._is_installed("docker") & \
```
The check if docker/systemd is installed should be done before we check if magmad is running with docker/systemd - and if docker/systemd is not installed then the follow up check should not be executed..

MoritzThomasHuebner replied:
I changed the `&` to `and`, this way the second statement will not be evaluated if the first one is `False`, e.g.:

```
>>> def func1():
...    print("func1")
...    return False
... 
>>> def func2():
...    print("func2")
...    return False
... 
>>> a = func1() and func2()
func1
>>> a = func1() & func2()
func1
func2
>>> 

```

VinashakAnkitAman replied:
@MoritzThomasHuebner I guess the statement of @nstng here means that you need to move `self._is_installed("docker")` and `self._is_installed("systemctl")` checks before line number 934 (before actually triggering shell commands in line numbers 935 and 938 using `exec_command_capture_output`). Or else if you want to move complete check in one like like you did now, you have to trigger shell commands together like below (but this makes command more complex and hence the old code looks simpler here):
```
docker_magmad_running = \
    self._is_installed("docker") and \
    (self.exec_command_capture_output( \
        "docker ps --filter 'name=magmad' --format '{{.Names}}'", \
    ).stdout.decode("utf-8").strip('\n') == 'magmad')
```

MoritzThomasHuebner replied:
Sorry, my mistake, it should be fixed now. However, I do have some issues with my VMs right now, so I can't properly test it at the moment.

nstng replied:
Structure now lgtm. Tested for `systemctl` (with running `magmad` and not running `magmadx`), output and behavior looks good!

MoritzThomasHuebner replied:
I tested with docker as well, and it behaves as expected.

VinashakAnkitAman commented on lte/gateway/python/integ_tests/s1aptests/s1ap_utils.py:
```diff
@@ -906,35 +903,43 @@ def exec_command_output(self, command):
+            e.g. 'sed -i \'s/config1/config2/g\' /etc/magma/mme.yml'
+
+        Returns:
+            output of command execution
```
How is this function `exec_command_capture_output` different from `exec_command_output`? As per the description both these functions return the same thing, i.e, command's output

MoritzThomasHuebner replied:
I added type hints and additional documentation. The difference is that this one does not raise an exception if the executed command returns a non-zero output.

VinashakAnkitAman replied:
Nice. Thanks for adding the documentation. Now the comment makes the need of this function clear.

MoritzThomasHuebner committed:
fix(lte): Improve changes in previous commits

- replace `&` with `and` in magmad running checks so that second condition is only evaluated if the first suceeds
- Improved docstring of `exec_command_capture_output`
- Added type hints to `exec_command`, `exec_command_output`, and `exec_command_capture_output` to highlight how these methods are different

Signed-off-by: Moritz Huebner <moritz.huebner@tngtech.com>

MoritzThomasHuebner committed:
fix(lte): Improve changes in previous commits

- Perform sequential checks for installation and if magma is running
- Remove exec_command_capture_output again as it is now unused

Signed-off-by: Moritz Huebner <moritz.huebner@tngtech.com>

MoritzThomasHuebner committed:
Revert "fix(lte): Improve changes in previous commits"

This reverts commit 2bb7908883b1ee396504264835dc60e3e4b335ac.

Signed-off-by: Moritz Huebner <moritz.huebner@tngtech.com>

MoritzThomasHuebner committed:
fix(lte): Properly moved installation checks first

Signed-off-by: Moritz Huebner <moritz.huebner@tngtech.com>

VinashakAnkitAman reviewed (approved):
LGTM. Please rebase the code as some ci check is failing.
